lib/client/client.go

```diff
@@ -548,58 +550,221 @@ func (client *NodeClient) scp(scpCommand scp.Command, shellCmd string, errWriter
+	}
+}
+
+func readByte(reader io.Reader) (byte, error) {
+	buf := []byte{0}
+	_, err := io.ReadFull(reader, buf)
+	return buf[0], err
+}
+
+const (
+	socks4Version                 byte = 0x04
+	socks5Version                 byte = 0x05
+	socks5Reserved                byte = 0x00
+	socks5AuthNotRequired         byte = 0x00
+	socks5AuthNoAcceptableMethods byte = 0xFF
+	socks5CommandConnect          byte = 0x01
+	socks5AddressTypeIPv4         byte = 0x01
+	socks5AddressTypeDomainName   byte = 0x03
+	socks5AddressTypeIPv6         byte = 0x04
+	socks5Succeeded               byte = 0x00
+)
+
+func socks5ProxyAuthenticate(incoming net.Conn) error {
+	nmethods, err := readByte(incoming)
+	if err != nil {
+		return err
+	}
+
+	chosenMethod := socks5AuthNoAcceptableMethods
+	for i := byte(0); i < nmethods; i++ {
+		method, err := readByte(incoming)
+		if err != nil {
+			return err
+		}
+		if method == socks5AuthNotRequired {
+			chosenMethod = socks5AuthNotRequired
+		}
+	}
+
+	_, err = incoming.Write([]byte{socks5Version, chosenMethod})
+	if err != nil {
+		return err
+	}
+
+	if chosenMethod == socks5AuthNoAcceptableMethods {
+		return errors.New("Unable to find suitable authentication method")
+	}
+
+	return nil
+}
+
+func socks5ProxyConnectRequest(incoming net.Conn) (remoteAddr string, err error) {
+	header := make([]byte, 4)
+	_, err = io.ReadFull(incoming, header)
+	if err != nil {
+		return
+	}
+	if !bytes.Equal(header[0:3], []byte{socks5Version, socks5CommandConnect, socks5Reserved}) {
+		err = errors.New("only connect command is supported for SOCKS5")
+		return
+	}
+
+	var ip net.IP
+	var remoteHost string
+	switch header[3] {
+	case socks5AddressTypeIPv4:
+		ip = make([]byte, net.IPv4len)
+	case socks5AddressTypeIPv6:
+		ip = make([]byte, net.IPv6len)
+	case socks5AddressTypeDomainName:
+		var domainNameLen byte
+		domainNameLen, err = readByte(incoming)
+		if err != nil {
+			return
+		}
+		remoteAddrBuf := make([]byte, domainNameLen)
+		_, err = io.ReadFull(incoming, remoteAddrBuf)
 		if err != nil {
-			log.Errorf("Failed to connect to node %v", remoteAddr)
 			return
 		}
-		defer conn.Close()
-		// start proxying:
-		doneC := make(chan interface{}, 2)
-		go func() {
-			io.Copy(incoming, conn)
-			doneC <- true
-		}()
-		go func() {
-			io.Copy(conn, incoming)
-			doneC <- true
-		}()
-		<-doneC
-		<-doneC
-		log.Debugf("nodeClient.listenAndForward(%v -> %v) exited", incoming.RemoteAddr(), remoteAddr)
+		remoteHost = string(remoteAddrBuf)
+	default:
+		err = errors.New("Unsupported address type for SOCKS5 connect request")
+		return
+	}
+
+	if ip != nil {
+		// Still need to read the ip address
+		_, err = io.ReadFull(incoming, ip)
+		if err != nil {
+			return
+		}
+		remoteHost = ip.String()
+	}
+
+	var remotePort uint16
+	err = binary.Read(incoming, binary.BigEndian, &remotePort)
+	if err != nil {
+		return
+	}
+
+	// Send the same minimal response as openSSH does
+	response := make([]byte, 4+net.IPv4len+2)
```

Comment:
What about non-success replies? See https://tools.ietf.org/html/rfc1928#section-6

Reply:
Can you wrap this logic up in it's own function? Maybe `ReplyError()` and `ReplySuccess()` type of functions in `lib/utils/socks.go`?